### PR TITLE
Add reset to pipes

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: src/System.IO.Pipelines/IPipe.cs [hunk 1]
```diff
@@ -8,5 +8,6 @@ public interface IPipe
     {
         IPipeReader Reader { get; }
         IPipeWriter Writer { get; }
+        void Reset();
```
Comment (davidfowl) [inline src/System.IO.Pipelines/IPipe.cs]: Now this is hella hacky 😄 . Reset on the interface!

File: src/System.IO.Pipelines/Pipe.cs [hunk 2]
```diff
@@ -85,8 +85,15 @@ public Pipe(BufferPool pool, PipeOptions options = null)
...
             _writerScheduler = options.WriterScheduler ?? InlineScheduler.Default;
+            ResetState();
+        }
+
+        private void ResetState()
+        {
             _readerAwaitable = new PipeAwaitable(completed: false);
             _writerAwaitable = new PipeAwaitable(completed: true);
+            _readerCompletion = default(PipeCompletion);
+            _writerCompletion = default(PipeCompletion);
```
Comment (davidfowl) [inline src/System.IO.Pipelines/Pipe.cs]: What about all of the other fields? Who resets length?

File: src/System.IO.Pipelines/IPipe.cs (hunk 1 above)
Comment (pakrym) [inline src/System.IO.Pipelines/IPipe.cs]: Nah, it's nice, owner of entire pipe should reset.

File: src/System.IO.Pipelines/Pipe.cs (hunk 2 above)
Comment (pakrym) [inline src/System.IO.Pipelines/Pipe.cs]: Good catch, would add a test.

File: src/System.IO.Pipelines/IPipe.cs (hunk 1 above)
Comment (pakrym) [inline src/System.IO.Pipelines/IPipe.cs]: We can do TryReset() or throw

**Pull request merged**